src/gui/sharedialog.cpp

```diff
@@ -0,0 +1,308 @@
+{
+    _path = path;
+    ShareDialog::getShares();
+}
+
+ShareDialog::~ShareDialog()
+{
+    delete _ui;
+}
+
+void ShareDialog::slotPasswordReturnPressed()
+{
+    ShareDialog::setPassword(_ui->lineEdit_password->text());
+    _ui->lineEdit_password->setPlaceholderText("Password Protected");
+    _ui->lineEdit_password->setText("");
+}
+
+void ShareDialog::setPassword(QString password)
+{
+    _ui->labelPasswordSpinner->show();
+    QUrl url = Account::concatUrlPath(AccountManager::instance()->account()->url(), QString("ocs/v1.php/apps/files_sharing/api/v1/shares/").append(QString::number(_public_share_id)));
+    QUrl postData;
+    QList<QPair<QString, QString> > getParams;
+    QList<QPair<QString, QString> > postParams;
+    getParams.append(qMakePair(QString::fromLatin1("format"), QString::fromLatin1("json")));
+    postParams.append(qMakePair(QString::fromLatin1("password"), password));
+    url.setQueryItems(getParams);
+    postData.setQueryItems(postParams);
+    OcsShareJob *job = new OcsShareJob("PUT", url, postData, AccountManager::instance()->account(), this);
+    connect(job, SIGNAL(jobFinished(QString)), this, SLOT(slotPasswordSet(QString)));
+    job->start();
+}
+
+void ShareDialog::slotPasswordSet(QString reply)
+{
+    _ui->labelPasswordSpinner->hide();
+}
+
+void ShareDialog::getShares()
+{
+    this->setWindowTitle(QString("Sharing ").append(_path));
+    QUrl url = Account::concatUrlPath(AccountManager::instance()->account()->url(), QLatin1String("ocs/v1.php/apps/files_sharing/api/v1/shares"));
+    QList<QPair<QString, QString> > params;
+    params.append(qMakePair(QString::fromLatin1("format"), QString::fromLatin1("json")));
+    params.append(qMakePair(QString::fromLatin1("path"), _path));
+    url.setQueryItems(params);
+    OcsShareJob *job = new OcsShareJob("GET", url, QUrl(), AccountManager::instance()->account(), this);
+    connect(job, SIGNAL(jobFinished(QString)), this, SLOT(slotSharesFetched(QString)));
+    job->start();
+}
+
+void ShareDialog::slotSharesFetched(QString reply)
+{
+    bool success = false;
+    QVariantMap json = QtJson::parse(reply, success).toMap();
+    ShareDialog::_shares = json.value("ocs").toMap().values("data")[0].toList();
+    for(int i = 0; i < ShareDialog::_shares.count(); i++)
+    {
+        QVariantMap data = ShareDialog::_shares[i].toMap();
+
+        if (data.value("share_type").toInt() == SHARETYPE_PUBLIC)
+        {
+            _public_share_id = data.value("id").toULongLong();
+
+            _ui->lineEdit_shareLink->show();
+            _ui->pushButton_copy->show();
+            _ui->checkBox_password->show();
+            _ui->checkBox_expire->show();
+            _ui->pushButton_copy->show();
+            _ui->checkBox_shareLink->setChecked(true);
+
+            if (data.value("share_with").isValid())
+            {
+                _ui->checkBox_password->setChecked(true);
+                _ui->lineEdit_password->setText("********");
```

Comment:
My proposal would be to keep the dummy stars, but to clear the line edit when it receives focus. Then, when no new password has been set, but the respective checkbox also has not been disabled, don't make any effective change, but display the dummy `****` again once you tab out.